mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp

```diff
@@ -1069,39 +1069,6 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
   return {};
 }
 
-namespace {
-/// Fold dim of a memref reshape operation to a load into the reshape's shape
-/// operand.
-struct DimOfMemRefReshape : public OpRewritePattern<DimOp> {
-  using OpRewritePattern<DimOp>::OpRewritePattern;
-
-  LogicalResult matchAndRewrite(DimOp dim,
-                                PatternRewriter &rewriter) const override {
-    auto reshape = dim.getSource().getDefiningOp<ReshapeOp>();
-
-    if (!reshape)
-      return failure();
-
-    // Place the load directly after the reshape to ensure that the shape memref
-    // was not mutated.
-    rewriter.setInsertionPointAfter(reshape);
-    Location loc = dim.getLoc();
-    Value load =
-        rewriter.create<LoadOp>(loc, reshape.getShape(), dim.getIndex());
```

Comment:
To avoid the dominance error: Instead of full `DominanceInfo` you could do a cheaper check here that:
* `dim.getIndex()` is defined in a parent block of `reshape`.
* Or: `dim.getIndex()` is defined in the same block as `reshape` but before `reshape`.

The annoying part is that `dim.getIndex()` could be a block argument, so there's one more special case. I've have into use cases where I needed this kind of check (defined in parent op or in the same block before a given op) many times, so we could add a such a helper function to `Dominance.h`.

Reply:
> Or: dim.getIndex() is defined in the same block as reshape but before reshape.

If you do that, then you can just move the insertion point to the later of the index or the reshape I think.

Reply:
>     * `dim.getIndex()` is defined in a parent block of `reshape`.
This will require recursing through all parent blocks of `reshape` -- is that cheap enough to be part of canonicalization? Should this be it's own pass itself instead of a canonicalization pattern?

Reply:
IMO it's cheap enough. You can do something along the lines of `dim.getIndex().getParentRegion()->isProperAncestor(reshape->getParentRegion())`.

Reply:
Oh great, that does what I was thinking. Still learning about the codebase. Thanks!

However, with the `tensor.dim` canonicalization, is there a requirement for canonicalizing `memref.dim` for the same pattern too?

Reply:
`tensor.dim` and `memref.dim` used to be the same operation and we split it into two separate ops at some point. One the one hand, op canonicalizations should be defined in the dialect of the op (i.e., splitting it up would make sense). On the other hand, in this case it means duplicating some code. Personally, I'd rather have two patterns, one in the tensor dialect and one in the memref dialect, so that the two dialects are decoupled as much as possible.